### DISCUSSION TITLE
Seeing duplicated commit trailers when using squash and merge in single commit PRs when committed

### DISCUSSION THREAD

**Discussion by tristan957** [community/community#29733] [Pull Requests]
Here is an example PR: https://github.com/hse-project/hse/pull/393

Here is the commit on the default branch: https://github.com/hse-project/hse/commit/1a905a79ce264498547434897f1f0b814faf4c33

You can see two `Signed-off-by`s in the commit. This has only began within the last week or so.

**Comment by willsmythe** [comment]
Thanks for reporting. We'll investigate.

**Reply by tristan957** [reply]
Thanks Will for taking a look. Much appreciated.

**Comment by jennyli-z** [comment]
we are facing the same issue. Any progress?

**Comment by tristan957** [comment]
The longer this goes on the worse the commit history goes for my projects. I can only warn my teammates so many times before it leads to nagging.

**Reply by tristan957** [reply]
This has now been happening for well over a month. @willsmythe could we have an update?

**Comment by tristan957** [comment]
I can confirm that this doesn't seem to occur for `Co-Authored-By` trailers

**Reply by gordey4doronin** [reply]
For me actually also happens for `co-authored-by`.. 😕 
https://github.com/gordey4doronin/co-authored-by/commit/08afa72ecb0e5f4b9d0132bc074ba0c31a2b58c7

<img width="939" alt="Screenshot 2022-11-21 at 17 50 33" src="https://user-images.githubusercontent.com/2134746/203113842-c2946a8a-c50e-49f7-b4ff-ade1d82c2c8d.png">

**Comment by tristan957** [comment]
Are there any updates after a month and a half? This bug is killing Squash and Merge as a legitimate strategy if you require DCO.

**Reply by tristan957** [reply]
2 months of a broken squash and merge strategy.

**Comment by gordey4doronin** [comment]
@tristan957 For the single-commit case it's clear that the duplicates should be removed. 👍 

What is your expectation about multiple commits squash-merge cases? For example this one https://github.com/hse-project/hse/commit/2c320ce22640dc6677794365e2267a74e114e266.

Every commit has `Signed-off-by` trailer there (red).
And the final squash-merge-commit combines all the trailers fro every commit as the final trailer (green).
And in the end it also looks odd.

<img width="1259" alt="Screenshot 2022-11-21 at 17 40 16" src="https://user-images.githubusercontent.com/2134746/203111198-fa7efc8f-6184-4e4f-89fd-cac9d918e70c.png">

<img width="801" alt="Screenshot 2022-11-21 at 17 37 23" src="https://user-images.githubusercontent.com/2134746/203111236-b8bf101c-c418-487a-bc42-6073bb641edb.png">

**Reply by tristan957** [reply]
Ideally it would be nice if all the extra `Signed-off-by`s could be removed. I try to get my teammates to edit their commits, but that has proven pretty difficult.

I could see either behavior being valid though.

**Reply by gordey4doronin** [reply]
@tristan957 We rolled the fix for @hse-project organization (all repos owned by it).
* For single-commit scenarios the duplicates should not happen from now on
* For multiple-commit scenarios - there will be a separator added

Feel free give it a try and let us know what you think.
Happy upcoming holidays!

**Reply by fruffy** [reply]
Hi @gordey4doronin, what was the fix for this? We are seeing a similar problem for our repository which requires DCO: 

https://github.com/p4lang/p4c/commit/0ccd7c6a6b9b37a3fcdc4c67845006c88533d348

![image](https://github.com/user-attachments/assets/ccb6e999-9b8c-4e78-a6ed-5b9841f37af2)

**Reply by gordey4doronin** [reply]
@fruffy The commit you linked already contains the fix we made. Look at the bottom, it has the separator:
```
---------

Signed-off-by: Anton Korobeynikov <anton@korobeynikov.info>
```

Basically we take all the trailers, de-duplicate them, and add de-duplicated list after the separator. 
In your case all the commits have been made by Anton Korobeynikov, and this is the resulting trailer added below the `---------` separator.

**Reply by fruffy** [reply]
What confuses me is that there is no deduplication happening here. We have about a dozen 
`Signed-off-by: Anton Korobeynikov <anton@korobeynikov.info>` in this commit.

**Reply by gordey4doronin** [reply]
De-duplication happens for the final trailer only, i.e. you don't receive dozens of `Signed-off-by: Anton Korobeynikov <anton@korobeynikov.info>` in the resulting trailer. And the resulting trailer is what matters for parsing all the contributions/coauthors etc.

In other words: we don't modify the original messages of the commits, and I don't think we should.
